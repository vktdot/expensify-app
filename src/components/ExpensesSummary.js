import React from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import numeral from 'numeral'
import selectExpensesTotal from '../selectors/expenses-total'
import  getVisibleExpenses from '../selectors/expenses'

export const ExpensesSummary = (props) => {
    const expenseWord = props.expenseCount === 1? 'expense': 'expenses';
    const formattedTotal = numeral(props.expensesTotal/100).format('$0,0.00')
    return(
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">Viewing <span> {props.expenseCount} </span>
                 {expenseWord} totalling <span> {formattedTotal} </span></h1>
                 <div className="page-header__actions">
                    <Link className="button" to="/create">Add Expense</Link>
                 </div>
            </div>
        </div>
        )
    
}

const MapStateToProps = (state) => {
    const visibleExpenses = getVisibleExpenses(state.expenses, state.filters);
    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: selectExpensesTotal(visibleExpenses)
    }
}
    
    
export default connect(MapStateToProps)(ExpensesSummary)