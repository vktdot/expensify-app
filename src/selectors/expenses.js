import moment from 'moment';

export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true
        const endDateMatch = endDate? endDate.isSameOrAfter(createdAtMoment, 'day') : true
        const textMatch = typeof text!== 'string' || expense.description.toLowerCase().includes(text.toLowerCase())  ;
        return startDateMatch && endDateMatch && textMatch;
    }).sort(function(a,b){
        if(sortBy === 'date'){
            return b.createdAt - a.createdAt
        }
        else{
            return b.amount - a.amount
        }
    })
}