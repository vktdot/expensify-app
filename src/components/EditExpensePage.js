import React from 'react';
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { editExpense, startRemoveExpense, startEditExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    onSubmit= (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    }
    onRemove = () =>{
        this.props.startRemoveExpense(this.props.expense.id);
        this.props.history.push('/')
    } 
    render(){
        return(
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm 
                    onSubmit = {this.onSubmit}
                    expense = {this.props.expense}
                    />
                    <button className = "button button--secondary" onClick = {this.onRemove}>
                        Remove Expense
                     </button>
                </div>
            </div>
        )
    }
}

// const EditExpensePage = (props) => {
//         return(
//             <div>
//                 <ExpenseForm 
//                   expense = {props.expense}
//                   onSubmit={(expense) => {
//                       props.dispatch(editExpense(expense.id, expense));
//                       props.history.push('/');
//                   }}
//                 />
//                 <button onClick = {() =>{
//                     props.dispatch(removeExpense(props.expense.id));
//                     props.history.push('/')}}  >
//                   Remove
//                 </button>
//             </div>
//         )
// }

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
})
    
const mapDispatchToProps = (dispatch) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (expense) => dispatch(startRemoveExpense(expense))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);