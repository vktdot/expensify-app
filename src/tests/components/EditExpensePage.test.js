import React from 'react'
import { shallow } from 'enzyme'
import { EditExpensePage } from '../../components/EditExpensePage'
import expenses from '../fixtures/expenses';

test('should render edit expense page', () => {
    const editExpense = jest.fn()
    const removeExpense = jest.fn()
    const wrapper = shallow(<EditExpensePage editExpense={editExpense} 
        removeExpense = {removeExpense}
        history = {history}
        expense = {expenses[2]}
        />)
    expect(wrapper).toMatchSnapshot();
})

test('should handle edit expense', () => {
    const editExpense = jest.fn()
    const history = {push: jest.fn()}
    const wrapper = shallow(<EditExpensePage editExpense={editExpense} 
        history = {history}
        expense = {expenses[2]}
        />)
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2])
    expect(history.push).toHaveBeenLastCalledWith('/')
    expect(editExpense).toHaveBeenLastCalledWith(expenses[2].id, expenses[2])
})

test('should handle remove expense', () =>{
    const removeExpense = jest.fn()
    const history = {push: jest.fn()}
    const wrapper = shallow(<EditExpensePage removeExpense = {removeExpense}
        history = {history}
        expense = {expenses[2]}
        />)
    wrapper.find('button').simulate('Click');
    expect(history.push).toHaveBeenLastCalledWith('/')
    expect(removeExpense).toHaveBeenLastCalledWith(expenses[2].id)
})