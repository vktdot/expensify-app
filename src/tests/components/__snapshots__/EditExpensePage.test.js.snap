// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render edit expense page 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 25000,
        "createdAt": -345600000,
        "description": "Shoes",
        "id": "3",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
     Remove 
  </button>
</div>
`;
